### PR TITLE
WIP: ✨ Support AWS multitenancy

### LINKED ISSUE #1552: AWS account multitenancy using a single instance of the controller
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
At present, CAPA relies on the default authentication providers within the AWS SDK.
The principle is described in the Java SDK at https://docs.aws.amazon.com/sdk-for-java/v2/developer-guide/credentials.html, with the Go SDK supporting the following chain:

1. If `AWS_WEB_IDENTITY_TOKEN_FILE` and `AWS_IAM_ROLE_ARN` env vars are both set, use refreshable role assumption via OIDC federation (v1alpha3 onwards).
2. Environment Variables (`AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, etc...)
3. Shared credentials file specified by `AWS_SHARED_CREDENTIALS_FILE` environment variable (defaulting to ~/.aws/config)
4. ECS refreshable credentials from the URL specified in the `AWS_CONTAINER_CREDENTIALS_FULL_URI` env var, using the env var `AWS_CONTAINER_AUTHORIZATION_TOKEN` as the authorisation header
5. ECS refreshable credentials from `http://169.254.170.2/${AWS_CONTAINER_CREDENTIALS_RELATIVE_URI}`
6. EC2 metadata

In addition, the [shared credentials file supports a wide variety of combinations](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html).

This provides maximum flexibility in the provision of credentials. In practice, this is implemented in CAPA by not specifying any credential parameters.

However, this means that a separate instance of CAPA must exist for each account/role/user that CAPA is meant to manage a cluster under, which may affect the deployment of conversion & defaulting webhooks as well as placing a scalability limit on running CAPA.

If possible, CAPA could support multi-tenancy within a single instance whilst preserving flexibility in terms of the authentication methods on offer.

We can break the above authentication methods into the following categories:

1. Forms of static credential provisioning
     * Env vars and static credentials encoded in the shared configuration file
2. Dynamic retrieval of credentials from a URL
     * ECS & EC2 IMDS metadata
3. Role assumption
     * Faked EC2 IMDS metadata provided by a system like Kiam
     * OIDC federation
     * Shared configuration file specifying role assumption in a profile with a static or dynamic EC2 metadata source

2 is specific to processes running on EC2 hosts or on ECS, and is not relevant, leaving 1 & 3 as methods which should be supportable for multi-tenancy within a single process.

From a Kubernetes perspective, any supply of credentials should be secured - if it comes from an API server resource, that API server resource must have appropriate RBAC (use of secret references, restricted CRD etc...).

### LINKED ISSUE #1696: Accept provided VPC infrastructure with only a private subnet
/kind feature

As a user, I may want to provide CAPA with a VPC that has no associated public subnets or internet gateways. It is unclear whether CAPA would support this today, even with an `internal` load balancing scheme for the apiserver.

**Anything else you would like to add:**

The current subnet code may issue an error when reconciling an unmanaged VPC with no public subnets: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/99349b8abeb34123b2e13d025946a2e66d6d6b40/pkg/cloud/services/ec2/subnets.go#L77-L87

I can't find anywhere else we mandate public subnets outside of points-of-use, and it looks like we may not even hit that check if the VPC has more than one subnet in it. More testing would be required to validate CAPA's behavior in this scenario.

### PR DESCRIPTION
**Pull request by andrewmyhre**
(Rebased for v0.6.0 / main)

**What this PR does / why we need it**: Allows the use of multiple credentials from as single deployment of CAPA. Static credentials and IAM roles are both support. These can be provided as custom resources along with a CAPx spec.

Fixes #1552 

TODO:

- [ ] Wire up namespace selector (@randomvariable )
- [ ] Tests for provider hashing, maybe.
- [ ] Add a way to test via e2e, even if we can't do it directly in Prow?

### CODE REVIEW

**Comment by randomvariable** [comment]
Not sure the rebase has gone correctly? Seems to include the multi-tenancy stuff as well.

**Comment by andrewmyhre** [comment]
> Not sure the rebase has gone correctly? Seems to include the multi-tenancy stuff as well.

I copied the title and description from the wrong PR :)
This is only the multi-tenancy work.

**Comment by randomvariable** [comment]
Cool, thanks. I'm going to be out until Tuesday, but will start reviewing then.

**Comment by randomvariable** [comment]
don't worry about rebasing just yet.

File: pkg/cloud/scope/principal.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	credentials *credentials.Credentials
+	// these are for tests :/
+	accessKeyID     string
+	secretAccessKey string
+	sessionToken    string
+}
+
+func (p *AWSStaticPrincipalTypeProvider) Hash() (string, error) {
+	var roleIdentityValue bytes.Buffer
+	err := gob.NewEncoder(&roleIdentityValue).Encode(p)
```
Comment (randomvariable) [inline pkg/cloud/scope/principal.go:74]: TIL: gob is a thing.

**Review by randomvariable — COMMENTED** [review]
Thanks for this @andrewmyhre . It's a great start.

There's obviously some bits we need to finish to complete the implementation, and would be willing to help in the effort.

Should we work on a common branch, either a feature branch in this repo or your repo, work on the bits together and then PR it as one?

File: test/e2e/e2e_suite_test.go [hunk 2]
```diff
@@ -30,6 +30,11 @@ import (
 	"testing"
 	"time"
 
+	"github.com/onsi/ginkgo/reporters"
+
+	"github.com/onsi/ginkgo/config"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+
```
[the remark this answers is not in the document]
Comment (andrewmyhre) [reply test/e2e/e2e_suite_test.go:37]: Yeah, not sure why Goland added these or why I didn't get an unused imports warning somewhere

File: pkg/cloud/services/elb/loadbalancer.go [hunk 3]
```diff
@@ -127,20 +127,20 @@ func (s *Service) DeleteLoadbalancers() error {
 
 	elbs, err := s.listOwnedELBs()
 	if err != nil {
-		return err
+		return errors.Errorf("Listing owned ELBs (1): %v", err)
```
[the remark this answers is not in the document]
Comment (andrewmyhre) [reply pkg/cloud/services/elb/loadbalancer.go:130]: Was intending to make this line consistent with other error handling.. boy scout rule. I can revert.

**Comment by randomvariable** [comment]
I'll do the rebase and add the namespace selector tomorrow.

File: config/manager/manager_image_patch.yaml [hunk 4]
```diff
@@ -8,5 +8,5 @@ spec:
     spec:
       containers:
       # Change the value of image field below to your controller image URL
-      - image: gcr.io/k8s-staging-cluster-api-aws/cluster-api-aws-controller:latest
+      - image: artifactory.cloud.capitalone.com/onekube/cluster-api-aws-controller-amd64:dev
```
Comment (MarcusNoble) [inline config/manager/manager_image_patch.yaml:11]:
```suggestion
      - image: gcr.io/k8s-staging-cluster-api-aws/cluster-api-aws-controller:latest
```

**Comment by randomvariable** [comment]
Completed in #2253 

Credit should be copied over to the other commits.

/close

**Pull request closed without merging**